### PR TITLE
[Step 2] Implement "small substs optimization" for substs of length 1 

### PR DESCRIPTION
**Pull request by csmoe**
Address #58310

### CODE REVIEW

File: src/librustc/infer/error_reporting/mod.rs [hunk 1]
```diff
@@ -798,7 +798,7 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                 if !has_default {
                     break;
                 }
-                if self.tcx.type_of(def_id).subst(self.tcx, substs) != actual {
+                if self.tcx.type_of(def_id).subst(self.tcx, &substs) != actual {
```
Comment (oli-obk) [inline src/librustc/infer/error_reporting/mod.rs]: I dislike all these additional references. Can we get rid of those via some trait impls or by changing the arguments of all the functions?

File: src/librustc/infer/opaque_types/mod.rs [hunk 2]
```diff
@@ -380,7 +380,7 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                         substs,
                         item_def_id: _,
                     }) => {
-                        for k in substs {
+                        for k in substs.iter() {
```
Comment (oli-obk) [inline src/librustc/infer/opaque_types/mod.rs]: Implement `IntoIterator` for `&SubstsRef`

File: src/librustc/ty/subst.rs [hunk 3]
```diff
@@ -384,9 +409,9 @@ impl<'tcx> TypeFoldable<'tcx> for SubstsRef<'tcx> {
         // If folding doesn't change the substs, it's faster to avoid
         // calling `mk_substs` and instead reuse the existing substs.
         if params[..] == self[..] {
-            self
+            *self
         } else {
-            folder.tcx().intern_substs(&params)
+            folder.tcx().intern_substs(&params).into()
```
Comment (oli-obk) [inline src/librustc/ty/subst.rs]: `intern_substs` should directly return a `SubstsRef`

**Comment by Centril** [comment]
Ping from triage, @csmoe, you have some rebasing + review comments to address.

File: src/librustc/ty/subst.rs (hunk 3 above)
Comment (csmoe) [reply src/librustc/ty/subst.rs]: @oli-obk 
As `_intern_substs` comes from [slice_interners!](https://github.com/rust-lang/rust/blob/d4a32d504a5aa49b951bfc70602a9615cb772acf/src/librustc/ty/context.rs#L2375-L2393), which returns a `List`, then should I implement a specific one for `SubstsRef`?

File: src/librustc/ty/print/mod.rs [hunk 4]
```diff
@@ -199,7 +199,7 @@ pub trait Printer<'gcx: 'tcx, 'tcx>: Sized {
             }
         }).count();
 
-        &substs[own_params]
+        SubstsRef::from_slice(self.tcx(), &substs[own_params])
```
Comment (oli-obk) [inline src/librustc/ty/print/mod.rs]: why does this take `TyCtxt` as an argument? Is there a method that will just pass through the reference if it already has the appropriate lifetime?

File: src/librustc/ty/structural_impls.rs [hunk 5]
```diff
@@ -487,7 +487,7 @@ impl<'a, 'tcx> Lift<'tcx> for ty::ProjectionTy<'a> {
         tcx.lift(&self.substs).map(|substs| {
             ty::ProjectionTy {
                 item_def_id: self.item_def_id,
-                substs,
+                substs: substs.clone(),
```
Comment (oli-obk) [inline src/librustc/ty/structural_impls.rs]: why do you need to clone here?

Comment (oli-obk) [reply src/librustc/ty/structural_impls.rs]: and all the other cases in this file?

File: src/librustc/ty/sty.rs [hunk 6]
```diff
@@ -322,31 +322,31 @@ pub struct ClosureSubsts<'tcx> {
...
     /// ordering.
-    fn split(self, def_id: DefId, tcx: TyCtxt<'_, '_, '_>) -> SplitClosureSubsts<'tcx> {
+    fn split(self, def_id: DefId, tcx: TyCtxt<'_, '_, 'tcx>) -> SplitClosureSubsts<'tcx> {
         let generics = tcx.generics_of(def_id);
         let parent_len = generics.parent_count;
         SplitClosureSubsts {
             closure_kind_ty: self.substs.type_at(parent_len),
             closure_sig_ty: self.substs.type_at(parent_len + 1),
-            upvar_kinds: &self.substs[parent_len + 2..],
+            upvar_kinds: SubstsRef::from_slice(tcx, &self.substs[parent_len + 2..]),
```
Comment (oli-obk) [inline src/librustc/ty/sty.rs]: same here, `tcx` isn't really necessary, the lifetimes should already fit

File: src/librustc/ty/sty.rs [hunk 7]
```diff
@@ -322,31 +322,31 @@ pub struct ClosureSubsts<'tcx> {
...
-    pub fn upvar_tys(self, def_id: DefId, tcx: TyCtxt<'_, '_, '_>) ->
+    pub fn upvar_tys(&self, def_id: DefId, tcx: TyCtxt<'_, '_, 'tcx>) ->
         impl Iterator<Item=Ty<'tcx>> + 'tcx
     {
-        let SplitClosureSubsts { upvar_kinds, .. } = self.split(def_id, tcx);
+        let SplitClosureSubsts { upvar_kinds, .. } = &self.split(def_id, tcx);
         upvar_kinds.iter().map(|t| {
             if let UnpackedKind::Type(ty) = t.unpack() {
-                ty
+                ty.clone()
```
Comment (oli-obk) [inline src/librustc/ty/sty.rs]: This copy seems unnecessary to me, it's `Ty`

File: src/librustc/ty/sty.rs [hunk 8]
```diff
@@ -421,31 +421,31 @@ impl<'tcx> GeneratorSubsts<'tcx> {
...
 
     #[inline]
-    pub fn upvar_tys(self, def_id: DefId, tcx: TyCtxt<'_, '_, '_>) ->
+    pub fn upvar_tys(&self, def_id: DefId, tcx: TyCtxt<'_, '_, 'tcx>) ->
         impl Iterator<Item=Ty<'tcx>> + 'tcx
     {
-        let SplitGeneratorSubsts { upvar_kinds, .. } = self.split(def_id, tcx);
-        upvar_kinds.iter().map(|t| {
+        let SplitGeneratorSubsts { upvar_kinds, .. } = &self.split(def_id, tcx);
+        upvar_kinds.clone().iter().map(|t| {
```
Comment (oli-obk) [inline src/librustc/ty/sty.rs]: why are you cloning the iterator? it's not used again afterwards

File: src/librustc/ty/subst.rs (hunk 3 above)
Comment (oli-obk) [reply src/librustc/ty/subst.rs]: You should be able to just modify that method, too. I don't think it's used elsewhere.

**Comment by jonas-schievink** [comment]
Triage ping @csmoe, there's merge conflicts waiting to be resolved, and also a few review comments

**Comment by JohnCSimon** [comment]
Triage ping @csmoe, there's merge conflicts waiting to be resolved, and also a few review comments

**Pull request closed without merging**